torch/distributed/_tools/sac_estimator.py

```diff
@@ -0,0 +1,995 @@
+        curr_idx, output_ids, mod_inplace_info = self._get_inplace_metadata(
+            func, out_storages
+        )
+        # 4. Determine if the function is in-place, random-op or a view-like
+        is_view_like = is_view_fn(func) or is_inplace_view_fn(func)
+        is_rand_op = torch.Tag.nondeterministic_seeded in func.tags
+        if is_view_like:
+            nbytes = 0
+        # sdpa has non-deterministic seed, but might be deterministic
+        # if no dropout is applied
+        if func.overloadpacket.__name__ == "_scaled_dot_product_flash_attention":
+            is_rand_op = kwargs.get("dropout_p", 0) != 0
+        # 5. Create metadata information per active non-leaf module
+        for mod_fqn in self._mod_tracker.parents:
+            if mod_fqn in self._leaf_modules:
+                continue
+            acm = _SACMetadata(
+                func=func,
+                time_taken=op_time,
+                memory_used=nbytes,
+                curr_idx=curr_idx,
+                output_ids=output_ids,
+                inplace_info=mod_inplace_info[mod_fqn],
+                is_view_like=is_view_like,
+                is_rand_op=is_rand_op,
+            )
+            if acm_stats := self._sac_mod_metadata.get(mod_fqn, None):
+                acm_stats.sac_metadata.append(acm)
+            else:
+                assert (
+                    mod_fqn == "Global"
+                ), f"Module {mod_fqn} not found in AC Mod Stats"
+                self._sac_metadata.append(acm)
+
+        return out
+
+    def _get_greedy_order_meta(self, sac_stats: SACStats) -> SACGreedyOrderMeta:
+        # An inplace-op group is a set of inplace-ops that operate on the same underlying tensor storage.
+        # 1. inplace_op_groups: A dictionary from the top-most parent of inplace-ops to the inplace-ops in the group
+        #   The top-most op can itself be an inplace-op or can be a non-inplace op.
+        # 2. inplace_op_to_group_head: A dictionary that maps all the inplace-ops to their respective group heads.
+        inplace_op_groups: Dict[int, Set[int]] = {}
+        inplace_op_to_group_head: Dict[int, int] = dict(sac_stats.inplace_ops)
+
+        # Initialize inplace_op_groups using inplace_op_to_group_head
+        for op_idx, group_head_idx in inplace_op_to_group_head.items():
+            op_group = inplace_op_groups.setdefault(group_head_idx, {group_head_idx})
+            op_group.add(op_idx)
+
+        # Like inplace ops, all of the random ops in the function/module should all be either recomputed or saved
+        # as a group. This is because, they affect the ranom seed generator. If force_store_random is set True,
+        # all of the random ops will be stored by default. For easy of manageability, we store the top-most random op
+        # as the leader of the random_ops_group.
+        random_ops_group: Dict[int, Set[int]] = {}
+        random_group_head_idx = min(sac_stats.rand_ops)
```

Comment:
Trying to understand this on a small model like below, but run into an error here, which I think is because there is no random ops and `sac_stats.rand_ops` is empty.

```
  File "/data/users/xuanzh/pytorch/torch/distributed/_tools/sac_estimator.py", line 512, in _get_greedy_order_meta
    random_group_head_idx = min(sac_stats.rand_ops)
                            ^^^^^^^^^^^^^^^^^^^^^^^
ValueError: min() arg is an empty sequence
```

```
import torch
import torch.nn as nn
from torch.distributed._tools.sac_estimator import SACEstimator
from torch._subclasses.fake_tensor import FakeTensorMode

if __name__ == "__main__":
    class Foo(nn.Module):
        def __init__(self):
            super().__init__()
            self.fc1 = nn.Linear(5, 10)
            self.relu = nn.ReLU()

        def forward(self, x):
            x = self.fc1(x)
            x = self.relu(x)
            return x

    with FakeTensorMode():
        with torch.device("cuda"):
            model = Foo()
        x = torch.rand((10, 5), device="cuda")

        sac_estimator = SACEstimator()
        with sac_estimator(estimate_mode_type="operator-level-cost-model"):
            loss = model(x).sum()
        loss.backward()
            
 ````

Reply:
Shall we add some test cases for simple models and check the SACStats collected.

Reply:
discussed offline and will incorporate the fix in https://github.com/pytorch/pytorch/pull/137908/